Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CertificateProvider.kt

```diff
@@ -2,28 +2,74 @@ package de.rki.coronawarnapp.covidcertificate.common.certificate
+import de.rki.coronawarnapp.covidcertificate.vaccination.core.VaccinatedPerson
+import de.rki.coronawarnapp.covidcertificate.vaccination.core.VaccinationCertificate
 import de.rki.coronawarnapp.covidcertificate.vaccination.core.repository.VaccinationRepository
+import de.rki.coronawarnapp.util.coroutine.AppScope
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.flow.shareLatest
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.flow.distinctUntilChanged
 import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.plus
 import javax.inject.Inject
 
 @Reusable
 class CertificateProvider @Inject constructor(
-    private val vcRepo: VaccinationRepository,
-    private val tcRepo: TestCertificateRepository,
-    private val rcRepo: RecoveryCertificateRepository,
+    vcRepo: VaccinationRepository,
+    tcRepo: TestCertificateRepository,
+    rcRepo: RecoveryCertificateRepository,
+    @AppScope appScope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider
 ) {
 
+    val certificateContainer: Flow<CertificateContainer> = combine(
+        rcRepo.certificates,
+        tcRepo.certificates,
+        vcRepo.vaccinationInfos
+    ) { recoveries, tests, vaccinations -> CertificateContainer(recoveries, tests, vaccinations) }
+        .conflate()
+        .distinctUntilChanged()
+        .shareLatest(scope = appScope + dispatcherProvider.IO)
+
     /**
      * Finds a [CwaCovidCertificate] by [CertificateContainerId]
      * @throws [Exception] if certificate not found
      */
     suspend fun findCertificate(containerId: CertificateContainerId): CwaCovidCertificate {
-        val certificates = rcRepo.cwaCertificates.first() +
-            tcRepo.cwaCertificates.first() +
-            vcRepo.cwaCertificates.first()
-
+        val certificates = certificateContainer.first().allCwaCertificates
         return certificates.find { it.containerId == containerId }!! // Must be a certificate
     }
+
+    data class CertificateContainer(
+        val recoveryCertificates: Set<RecoveryCertificateWrapper>,
+        val testCertificates: Set<TestCertificateWrapper>,
+        val vaccinationInfos: Set<VaccinatedPerson>
+    ) {
+
+        val recoveryCwaCertificates: Set<RecoveryCertificate> by lazy {
+            recoveryCertificates.map { it.recoveryCertificate }.toSet()
+        }
+
+        val testCwaCertificates: Set<TestCertificate> by lazy {
+            testCertificates.mapNotNull { it.testCertificate }.toSet()
+        }
+
+        val vaccinationCwaCertificates: Set<VaccinationCertificate> by lazy {
+            vaccinationInfos.flatMap { it.vaccinationCertificates }.toSet()
+        }
+
+        val allCwaCertificates by lazy {
+            recoveryCwaCertificates + testCwaCertificates + vaccinationCwaCertificates
+        }
```

Comment:
be careful with using `val` once a value is assigned it won't change , does it update its value when certificates  changes like adding / removing ?

Reply:
I don't get your point here. The container is immutable. It is not possible to add or remove certificates to/from the container. If the certificates in a repo change, a new container gets created..